### PR TITLE
Adding Ls - de Levie Pore _Finite Length Element

### PR DESCRIPTION
**Pull request by ravuru66**
This element describes the impedance of a pore with conductive and reactive pore walls. This is in contrast to Finite Length Warburg element where the pores are non-reactive (infinite polarization resistance).

Z(w) = (R/(gamma)^1/2) * coth((gamma)^1/2)

where Z is the impedance , R is the resistance, w is the angular frequency and
gamma = (1/A)+B*(jw)^phi

The following referenes give a detailed description of the electrochemical model:

    A.Lasia, "Impedance of porous electrodes", Modern Aspects of electrochemistry, "Modellingand Numerical Simulations," vol.43, p.67-138, M.Schlesinger, Ed.,Springer, 2009, IsBN: 978-0-387-49580-4.

    r De Levie, Adv. electrochem. Elecrochem. eng., 6(1967) 329.

### CODE REVIEW

**Review by ravuru66 — COMMENTED** [review]
kindly check the changes and merge the part of program

**Comment by ravuru66** [comment]
> 
> 
> kindly check the changes and merge the part of program

**Comment by ravuru66** [comment]
Hi @mdmurbach 

I added the snippet code for the impedance of "Ls-de Levie Pore - finite element" in "impedance.py/impedance/models/circuits/elements.py /" and removed all the PEP8 errors.

Later in the "impedance.py/impedance/tests/test_circuit_elements.py /" , added key  'Ls' and it values for the said frequencies in the 'correct_vals' dictionary. 

After committing , Error: "impedance/tests/test_circuit_elements.py:41:24: E999 Syntax Error: invalid syntax" has shown. Could you please look into this error and resolve it. And then merge the element in the source code.

If possible provide me with the solution for the error and the further steps in merging the element to the source code which would be helpful in the future.

Feel free to make changes in the code I have added to the files and let me know. so that I will not repeat the errors again.

Thanks in advance.

**Comment by BGerwe** [comment]
Hi @ravuru66,

For future reference, you can locally check for PEP8 errors by running ```flake8 impedance --exclude=*.ipynb_checkpoints``` and you can run unit tests locally with ```pytest --cov=impedance impedance/tests```. That way you don't have to wait ~1 min each time you want to check if the errors were fixed, and it keeps the commit history cleaner.

**Comment by ravuru66** [comment]
Hi @BGerwe 

Thank you for the information.  I will implement it from the next time.

**Comment by mdmurbach** [comment]
Hi @ravuru66 thanks for starting this PR to add de Levie's porous electrode element! 

I'm a little confused where the form of the equation for Gamma/Lambda you're using comes from (might just be a long week and I'm missing it in the paper...), but I see the code with something like

Gamma = 1/(A + B*(j*omega)^phi)

whereas the ["Impedance of Porous Electrodes" paper](https://www.sciencedirect.com/science/article/abs/pii/0022072895041775?via%3Dihub) has

image

It seems like maybe what's in the code is a "CPE" version of the same equation?

**Comment by ravuru66** [comment]
Hi @mdmurbach Thank you for asking the question.

The constants A, B and Phi in the equation of gamma/lambda are calculated by substituting specific solution resistance  (p), cylinder pore radius(r), length (l), specific charge transfer resistance of pore walls(Rct), and specific capacitance of pore walls(Cdl) which are different for different electrodes.

So the equation written in the code is the simplified one and  infer the same meaning proposed in the paper. 

The scribner Associates,Inc who are the developers of the ZView software use the same equation from where I referred the equation.

Image

**Comment by BGerwe** [comment]
> It seems like maybe what's in the code is a "CPE" version of the same equation?

image

Yeah, I think you're exactly right. But that distinction is important, since otherwise this element is mathematically equivalent to the [finite-length Gerischer](https://impedancepy.readthedocs.io/en/latest/circuit-elements.html#impedance.models.circuits.elements.Gs) element. The derivation is different (Lasia vs Green), but that's because the former considers aqueous systems, and the latter considers solid-state systems. Otherwise, you can rearrange parameters until they are identical.

So, I think it's worthwhile including this element since it is unique. However, it does raise the question of should we include elements that are mathematically equivalent but have different names and interpretations? I'm inclined towards yes since users are unlikely to spend their time doing algebra to see if we already have an element they need.

Another point to raise is the naming of this element. So far, we have used "s" as a modifier to indicate finite-length ([Gs](https://impedancepy.readthedocs.io/en/latest/circuit-elements.html#impedance.models.circuits.elements.Gs), [Ws](https://impedancepy.readthedocs.io/en/latest/circuit-elements.html#impedance.models.circuits.elements.Ws)) and "L" most often means inductor, so I would prefer avoiding "Ls". We already have "T" representing a [porous model](https://impedancepy.readthedocs.io/en/latest/circuit-elements.html#impedance.models.circuits.elements.T), and "a" as a modifier for dispersion ([La](https://impedancepy.readthedocs.io/en/latest/circuit-elements.html#impedance.models.circuits.elements.La)). So, I propose naming this element "Tas" to indicate it's a porous electrode model of finite length with dispersion. (We can do 3 character names now, right?) But I'm open to other suggestions.

File: impedance/models/circuits/elements.py [hunk 1]
```diff
@@ -268,6 +268,33 @@ def Gs(p, f):
...
 
 
+@element_metadata(num_params=4, units=['Ohm', '', '', '', ''])
+def Tas(p, f):
+    """ defines a Ls-de Levie Pore - Finite Length
+
+    Notes
+    ---------
+    .. math::
+        Z = \\frac{Z_0}{\\sqrt{gamma}\\tanh{\\sart{gamma}}}
```
Comment (mdmurbach) [inline impedance/models/circuits/elements.py:278]: It looks like this doc string could benefit from some clean up. A few things I notice:

- First line Ls should be something like "defines a Finite-Length de Levie Pore element as represented in [1]" to mirror the Gerisher and other more complex elements
- The math seems to have some formatting problems. You should be able to build the documentation locally by: `cd docs`, `make html`, open build/html/index.html
- I can't figure out how to push to your branch, but here's what I was playing around with

```
Notes
    ---------
    .. math::
        Z = \\frac{R_{\\Omega,p}}{\\sqrt{\\Lambda}\\tanh{\\sqrt{\\Lambda}}}

    where

    .. math::
        R_{\\Omega,p} = \\frac{\\rho l}{\\pi r^2},

    .. math::
        \\Lambda = \\frac{1}{A} + B (j \\omega)^\\phi,

    and

    .. math::
        A = \\frac{R_{ct}r}{2\\rho l^2} \\quad B = \\frac{2\\rho l^2 C_{dl}}{r}

    where :math:`R_{\\Omega,p}` = p[0] (Ohms) and
    :math:`A` = p[1], `B` = p[2], `\\phi` = p[3]

    [1] Lasia, A. Impedance of porous electrodes.
    Journal of Electroanalytical Chemistry 397, 27–33 (1995).
    `doi:10.1016/0022-0728(95)04177-5
    <https://doi.org/10.1016/0022-0728(95)04177-5>`_.
    """
```

**Review by mdmurbach — CHANGES_REQUESTED** [review]

**Comment by mdmurbach** [comment]
Big thanks to the both of you! Sorry it took so long for me to reply, but I added a few thoughts on the review.

Overall, it looks great, except for some small tweaks to tie the equation into the paper in the documentation string. 🚀

**Comment by ravuru66** [comment]
ValueError: Tas not in allowed elements (dict_keys(['Ws', 'L', 'Wo', 'p', 'G', 'K', 'CPE', 'T', 'La', 'R', 'Gs', 'C', 'W', 's', '__builtins__']))

while I am working on different PC it is throwing an above error for the element I have added in the dictionary.

could you please tell me the steps to solve the error

**Comment by mdmurbach** [comment]
You may need to (re)install the local version if you want the changes you've made to be available. Try running `pip install -e .` (note the period at the end) in the main `impedance.py/` directory (the one with `setup.py`).

**Comment by ravuru66** [comment]
I have downloaded the whole package with the changes that have been made in the code from the github. And copied all the directories into my impedance package with replacing exisiting files in it. Then ran the command what @mdmurbach mentioned. But still it is showing the same error , "ValueError: Tas not in allowed elements (dict_keys(['Ws', 'L', 'Wo', 'p', 'G', 'K', 'CPE', 'T', 'La', 'R', 'Gs', 'C', 'W', 's', 'builtins']))."

And code is working fine in the PC where I have made changes but it is not working in the new PC.

I am new to github. Please guide me step by step by which I can run the code with the modifications in the different PC. Thanks in advance for your guidance.

**Comment by ravuru66** [comment]
Hello @mdmurbach  @BGerwe 

I want to include transmissionLine model into impedance package locally for the same you can see the code below. I made changes in elements.py and test_circuit_elements.py. 

When executing the same code by calling the circuit with only one element 'Cl' it is showing the following error:

UnboundLocalError: local variable 'split' referenced before assignment

If circuit is named with 'R0-Cl'  then it is showing the following error:

ValueError: non-broadcastable output operand with shape (36,) doesn't match the broadcast shape (36,36)

could you help me in resolving this issue. Anyway I want a code to work the single element 'Cl'.

```
@element_metadata(num_params=7, units=['Ohm', '', '', '', '', '', ''])
def Cl(p, f):
       
    def r_ion(n, r_ion_0, b_ion):
        
        return r_ion_0 * np.exp(b_ion * n)

    def z(n, omega, r_ct_0, b_ct, c_dl_0, b_cd):
        
        r_ct_n = r_ct_0 * np.exp(b_ct * n)
        c_dl_n = c_dl_0 * np.exp(b_cd * n)
    
        return r_ct_n / (1 + 1j * omega * r_ct_n * c_dl_n) 

    def imp(n, omega, r_ion_0, b_ion, r_ct_0, b_ct, c_dl_0, b_cd):
        
        # omega = 2*np.pi*np.array(f)
        # n, r_ion_0, b_ion, r_ct_0, b_ct, c_dl_0, b_cd = p[0], p[1], p[2], p[3], p[4], p[5], p[6]
 
        if n == 0:
            return z(0, omega, r_ct_0, b_ct, c_dl_0, b_cd)
        else:
            return  1 / (1 / z(n, omega, r_ct_0, b_ct, c_dl_0, b_cd) + 1/(r_ion(n, r_ion_0, b_ion) + imp(n-1, omega, r_ion_0, b_ion, r_ct_0, b_ct, c_dl_0, b_cd)))


    def calc_EIS(n, omega, r_ion_0, b_ion, r_ct_0, b_ct, c_dl_0, b_cd):
        
        # omega = 2*np.pi*np.array(f)
        # n, r_ion_0, b_ion, r_ct_0, b_ct, c_dl_0, b_cd = p[0], p[1], p[2], p[3], p[4], p[5], p[6]
        n0 = n
        Z = []
        for freq in omega:
            Z.append(imp(n, omega, r_ion_0/n0, b_ion, r_ct_0*n0, b_ct, c_dl_0/n0, b_cd))
            
        return Z
    
    n, r_ion_0, b_ion, r_ct_0, b_ct, c_dl_0, b_cd = p[0], p[1], p[2], p[3], p[4], p[5], p[6]
    omega = 2*np.pi*np.array(f)

    
    return calc_EIS(n, omega, r_ion_0, b_ion, r_ct_0, b_ct, c_dl_0, b_cd)

```

**Pull request closed without merging**